### PR TITLE
IA-144 Implement get organizations in tenant

### PR DESCRIPTION
**Pull request by miguelTaningcoYoppworks**
Primary Reviewer for GetOrganizations: @AlexWeinstein92 
Primary Reviewer for overall Tenant Code quality: @dag-yoppworks 
Primary Reviewer if available: @reid-spencer 

### Issue:

- GetOrganizations command is needed. There is no projection here
- Tenant has some unnecessary code (the sealed value stuff) for match blocks
- Existing code always persisted events when queries aren't actually events

### Changes:

- implement GetOrganizations
- refactored tenant responses to either be TenantEventResponses or TenantDataResponses
- cleaned up code

### Testing:
- manual testing
- ran ci testing

### CODE REVIEW

**Review by dag-yoppworks — CHANGES_REQUESTED** [review]
This looks really good. I just have a few nitpicky minor changes to request.

File: tenant/src/main/protobuf/com/improving/app/tenant/domain/improving-app-tenant-domain.proto [hunk 1]
```diff
@@ -63,13 +63,24 @@ message InfoEdited {
   TenantInfo new_info = 4;
 }
 
+message GetOrganizations {
+  option (scalapb.message).extends = "com.improving.app.common.serialize.PBMsgSerializable";
+  com.improving.app.common.domain.TenantId tenant_id = 1;
+}
+
+message OrganizationData {
```
[the remark this answers is not in the document]
Comment (AlexWeinstein92) [reply tenant/src/main/protobuf/com/improving/app/tenant/domain/improving-app-tenant-domain.proto:71]: I think that's a good idea

File: tenant/src/main/scala/com/improving/app/tenant/domain/Tenant.scala [hunk 2]
```diff
@@ -43,63 +43,73 @@ object Tenant {
...
-          case TenantRequestMessage.SealedValue.SuspendTenantValue(value) => suspendTenant(establishedState, value)
-          case TenantRequestMessage.SealedValue.EditInfoValue(value) => editInfo(establishedState, value)
+        command.request match {
+          case _: EstablishTenant => Left(StateError("Tenant is already established"))
+          case x: ActivateTenant => activateTenant(establishedState, x)
+          case x: SuspendTenant => suspendTenant(establishedState, x)
+          case x: EditInfo => editInfo(establishedState, x)
+          case x: GetOrganizations => getOrganizations(x, Some(establishedState))
+          case _ => Left(StateError("Command is not supported"))
         }
```
Comment (AlexWeinstein92) [inline tenant/src/main/scala/com/improving/app/tenant/domain/Tenant.scala:62]: Why is there no handler for TerminatedTenantState?

Comment (miguelTaningcoYoppworks) [reply tenant/src/main/scala/com/improving/app/tenant/domain/Tenant.scala:62]: Not sure. @dag-yoppworks should know why, after his changes to Tenant in the previous PR.

File: tenant/src/main/scala/com/improving/app/tenant/domain/Tenant.scala [hunk 3]
```diff
@@ -43,63 +43,73 @@ object Tenant {
...
-  private val commandHandler: (TenantState, TenantCommand) => ReplyEffect[TenantEvent, TenantState] = { (state, command) =>
-    val result: Either[Error, TenantEvent] = state match {
+  private val commandHandler: (TenantState, TenantCommand) => ReplyEffect[TenantResponse, TenantState] = { (state, command) =>
+    val result: Either[Error, TenantResponse] = state match {
       case UninitializedTenant =>
-        command.request.asMessage.sealedValue match {
-          case TenantRequestMessage.SealedValue.EstablishTenantValue(value) => establishTenant(value)
+        command.request match {
+          case x: EstablishTenant => establishTenant(x)
+          case x: GetOrganizations => getOrganizations(x)
```
[the remark this answers is not in the document]
Comment (AlexWeinstein92) [reply tenant/src/main/scala/com/improving/app/tenant/domain/Tenant.scala:51]: True, I remember seeing that but didn't realize it applied to initialized state as well

File: tenant/src/main/protobuf/com/improving/app/tenant/domain/improving-app-tenant-domain.proto [hunk 4]
```diff
@@ -81,4 +92,27 @@ message TenantEvent {
...
+  }
+}
+
+message TenantEventResponse {
+  option (scalapb.message).extends = "com.improving.app.common.serialize.PBMsgSerializable";
+  TenantEvent tenant_event = 1;
+}
+
+message TenantDataResponse {
+  TenantData tenant_data = 1;
```
[the remark this answers is not in the document]
Comment (miguelTaningcoYoppworks) [reply tenant/src/main/protobuf/com/improving/app/tenant/domain/improving-app-tenant-domain.proto:109]: This cannot be done:

`--jvm_0_out: com.improving.app.tenant.domain.TenantResponse.tenant_data_value: sealed oneofs may not be a case within another sealed oneof`

File: tenant/src/main/scala/com/improving/app/tenant/domain/Tenant.scala [hunk 5]
```diff
@@ -43,63 +43,73 @@ object Tenant {
...
           case _ => Left(StateError("Tenant is not established"))
         }
       case establishedState: EstablishedTenantState =>
```
[the remark this answers is not in the document]
Comment (miguelTaningcoYoppworks) [reply tenant/src/main/scala/com/improving/app/tenant/domain/Tenant.scala:54]: went ahead and refactored this to deal with the active and suspended states in the commandHandler.

**Review by AlexWeinstein92 — DISMISSED** [review]
LGTM

File: tenant/src/main/scala/com/improving/app/tenant/domain/Tenant.scala (hunk 2 above)
Comment (AlexWeinstein92) [reply tenant/src/main/scala/com/improving/app/tenant/domain/Tenant.scala:74]: Terminated handler does not have to be included here, but if not please create another ticket. It should be small enough that including it here is not too bad, but up to you

Comment (miguelTaningcoYoppworks) [reply tenant/src/main/scala/com/improving/app/tenant/domain/Tenant.scala:74]: I'll create another ticket as to limit the scope of this PR. https://yoppworks.atlassian.net/browse/IA-169

File: tenant/src/main/scala/com/improving/app/tenant/domain/Tenant.scala [hunk 6]
```diff
@@ -17,7 +17,7 @@ import java.time.Instant
 object Tenant {
   val TypeKey = EntityTypeKey[TenantCommand]("Tenant")
 
-  case class TenantCommand(request: TenantRequest, replyTo: ActorRef[StatusReply[TenantEvent]])
+  case class TenantCommand(request: TenantRequest, replyTo: ActorRef[StatusReply[TenantResponse]])
```
[the remark this answers is not in the document]
Comment (dag-yoppworks) [reply tenant/src/main/scala/com/improving/app/tenant/domain/Tenant.scala:20]: I'd be okay with other nomenclature but this is not a Command. Agreed, this is implementation specific, not architectural.

**Review by dag-yoppworks — APPROVED** [review]
Thanks!

**Pull request merged**